presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java

```diff
@@ -0,0 +1,389 @@
+                    if (groupByColumns.size() != numGroupByClause) {
+                        throw new PinotException(PinotErrorCode.PINOT_UNEXPECTED_RESPONSE,
+                                Optional.of(psql),
+                                String.format("Expected %d gby columns but got %s instead from pinot", numGroupByClause, groupByColumns));
+                    }
+                    // group by aggregation
+                    for (int groupByIndex = 0; groupByIndex < metricValuesForEachGroup.size(); groupByIndex++) {
+                        JsonNode row = metricValuesForEachGroup.get(groupByIndex);
+                        JsonNode group = row.get("group");
+                        if (group == null || !group.isArray() || group.size() != numGroupByClause) {
+                            throw new PinotException(PinotErrorCode.PINOT_UNEXPECTED_RESPONSE,
+                                    Optional.of(psql),
+                                    String.format("Expected %d group by columns but got only a group of size %d (%s)", numGroupByClause, group.size(), group));
+                        }
+                        if (groupToValue == null) {
+                            singleAgg[0] = row.get("value").asText();
+                            setValuesForGroupby(blockBuilders, types, numGroupByClause, group, singleAgg);
+                            ++rowCount;
+                        }
+                        else {
+                            groupToValue.computeIfAbsent(group, (ignored) -> new String[aggResults.size()])[aggrIndex] = row.get("value").asText();
+                        }
+                    }
+                }
+                else {
+                    Preconditions.checkState(sawGroupByResult == null || !sawGroupByResult);
+                    sawGroupByResult = false;
+                    // simple aggregation
+                    // TODO: Validate that this is expected semantically
+                    Preconditions.checkState(numGroupByClause == 0, "Expected no group by columns in pinot");
+                    setValue(types.get(aggrIndex), blockBuilders.get(aggrIndex), result.get("value").asText());
+                    rowCount = 1;
+                }
+            }
+
+            if (groupToValue != null) {
+                Preconditions.checkState(rowCount == 0, "Row count shouldn't have changed from zero");
+                groupToValue.forEach((group, values) ->
+                        setValuesForGroupby(blockBuilders, types, numGroupByClause, group, values));
+                rowCount = groupToValue.size();
+            }
+        }
+        else if (selectionResults != null) {
+            JsonNode columns = selectionResults.get("columns");
+            JsonNode results = selectionResults.get("results");
+            if (columns == null || results == null || !columns.isArray() || !results.isArray() || columns.size() != blockBuilders.size()) {
```

Comment:
Use `requireNonNull` as opposed to direct null checks like this. Also this entire if statement can be converted to use `checkArgument()`. This is normally how we check for well-formed parameters.

Reply:
Actually no: I want to throw a specific exception type such that it can show up properly in my analytics. We care about the number of external errors and an IllegalState exception would be more of an internal error imho.

Reply:
I see, fair point!